cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+        new_path = f"static/{filename}"
+        _create_component_version_content_safely(
+            component_version_pk=component_version.pk,
+            content_pk=content_pk,
+            key=new_path
+        )
+    return component_version.publishable_entity_version
+
+
+def _deduplicate_container(
+    context: MigrationContext,
+    source_key: UsageKey,
+    container_type: ContainerType,
+    title: str,
+) -> LibraryContainerLocator:
+    if not source_key or not source_key.block_id:
+        raise ValueError(f"Invalid source_key: {source_key}")
+
+    log.debug(f"Processing deduplication for source key: {source_key}")
+
+    # Check if we already processed this block
+    if context.is_already_migrated(source_key):
+        log.debug(f"Block {source_key} already exists, reusing existing target")
+        existing_version = context.get_existing_target(source_key)
+
+        return LibraryContainerLocator(
+            context.target_library_key,
+            container_type.value,
+            existing_version.key
+        )
+
+    # Generate new unique block ID
+    log.debug(f"Creating new block for {source_key.block_id}")
+
+    base_slug = (
+        source_key.block_id
+        if context.preserve_url_slugs
+        else _slugify_source_usage_key(
+            source_key,
+            title,
+        )
+    )
+    unique_slug = _find_unique_slug(context, base_slug)
+
+    log.info(f"Created new block with slug: {unique_slug}")
+
+    return LibraryContainerLocator(
+        context.target_library_key,
+        container_type.value,
+        unique_slug
+    )
+
+def _deduplicate_block_id(
+    context: MigrationContext,
+    source_key: UsageKey,
+    component_type: str,
+    olx: str,
+) -> LibraryUsageLocatorV2:
+    """
+    Deduplicate a block_id by appending a unique identifier if necessary.
+
+    Args:
+        context: The migration context
+        source_key: The original usage key from the source
+        component_type: The component type string
+
+    Returns:
+        A unique LibraryUsageLocatorV2 for the target
+
+    Raises:
+        ValueError: If source_key is invalid
+    """
+    if not source_key or not source_key.block_id:
+        raise ValueError(f"Invalid source_key: {source_key}")
+
+    log.debug(f"Processing deduplication for source key: {source_key}")
+
+    # Check if we already processed this block
+    if context.is_already_migrated(source_key):
+        log.debug(f"Block {source_key} already exists, reusing existing target")
+        existing_target = context.get_existing_target(source_key)
+
+        # Parse the key more safely
+        try:
+            key_parts = existing_target.key.split(":")
+            if len(key_parts) < 2:
+                raise ValueError(f"Invalid key format: {existing_target.key}")
+            block_id = key_parts[-1]
+        except (AttributeError, IndexError) as e:
+            raise ValueError(f"Failed to parse existing key: {existing_target.key}") from e
```

Comment:
I'm unsure if we should be doing direct string manipulation of these keys. As you're familiar with, when we work with OpaqueKeys (like CourseKeys) it's best practice to use the objects methods (`course_key.run`, etc.) instead of assuming the key format. But entity keys are just strings, so I'm not sure how we're supposed to safely work with them. These entity keys are new to the platform and I'm not sure if we have an established best practice yet... let me do some research and come back to this comment.

Reply:
@cmltaWt0 OK, I think the safest thing to do here would be load the related component and look at its `local_key`, which is guaranteed to always contain the block id:

```suggestion
        block_id = existing_target.component.local_key
```

It adds up to one query for each existing block, but I think this is worth the robustness, and if we see performance issues we can address it in the future.